File: platformer.py

---

## 💻 Код на 7 языках

### 1. Python – `platformer.py`

```python
#!/usr/bin/env python3
# platformer.py - Минимальный платформер на Python

import os
import sys
import time
import json
import threading
import keyboard
from colorama import init, Fore, Style

init(autoreset=True)

# Константы
WIDTH = 40
HEIGHT = 12
GRAVITY = 0.3
JUMP_SPEED = -5.5
PLAYER = '@'
WALL = '#'
COIN = '$'
EXIT = 'X'
EMPTY = ' '
RECORD_FILE = 'platformer_record.json'

# Уровень (массив строк)
LEVEL = [
    '                                        ',
    '                                        ',
    '          #   #                         ',
    '   $      #   #     #   #              ',
    '   @      #   #  $  #   #   X          ',
    '########   ########   #############    ',
]

# Дополним уровень до HEIGHT строк (остальные пустые)
while len(LEVEL) < HEIGHT:
    LEVEL.append(' ' * WIDTH)
LEVEL = LEVEL[:HEIGHT]

class Platformer:
    def __init__(self):
        self.width = WIDTH
        self.height = HEIGHT
        self.level = [list(row.ljust(WIDTH)) for row in LEVEL]
        self.player_x = 0
        self.player_y = 0
        self.vx = 0
        self.vy = 0
        self.on_ground = False
        self.coins = 0
        self.total_coins = sum(row.count('$') for row in LEVEL)
        self.exit_x = 0
        self.exit_y = 0
        self.find_player_and_exit()
        self.game_over = False
        self.won = False
        self.paused = False
        self.running = True
        self.start_time = time.time()
        self.record = self.load_record()
        self.lock = threading.Lock()

    def find_player_and_exit(self):
        for y, row in enumerate(self.level):
            for x, ch in enumerate(row):
                if ch == PLAYER:
                    self.player_x, self.player_y = x, y
                    self.level[y][x] = EMPTY
                elif ch == EXIT:
                    self.exit_x, self.exit_y = x, y

    def load_record(self):
        try:
            with open(RECORD_FILE, 'r') as f:
                data = json.load(f)
                return data.get('record', None)
        except:
            return None

    def save_record(self, time_taken):
        with open(RECORD_FILE, 'w') as f:
            json.dump({'record': round(time_taken, 2)}, f)

    def jump(self):
        if self.on_ground and not self.game_over and not self.paused:
            self.vy = JUMP_SPEED
            self.on_ground = False

    def update_physics(self):
        # Гравитация
        self.vy += GRAVITY
        if self.vy > 10:
            self.vy = 10

        # Горизонтальное движение (с инерцией)
        self.player_x += self.vx

        # Проверка столкновений по X
        if self.collides(self.player_x, self.player_y):
            if self.vx > 0:
                self.player_x = self.get_left_collision(self.player_x, self.player_y)
            elif self.vx < 0:
                self.player_x = self.get_right_collision(self.player_x, self.player_y)
            self.vx = 0

        # Вертикальное движение
        self.player_y += self.vy

        # Проверка столкновений по Y
        if self.collides(self.player_x, self.player_y):
            if self.vy > 0:  # падение
                self.player_y = self.get_top_collision(self.player_x, self.player_y)
                self.vy = 0
                self.on_ground = True
            elif self.vy < 0:  # прыжок
                self.player_y = self.get_bottom_collision(self.player_x, self.player_y)
                self.vy = 0

        # Сбор монет
        if self.level[self.player_y][self.player_x] == COIN:
            self.level[self.player_y][self.player_x] = EMPTY
            self.coins += 1

        # Проверка выхода
        if self.player_x == self.exit_x and self.player_y == self.exit_y:
            self.won = True
            self.game_over = True
            elapsed = time.time() - self.start_time
            if self.record is None or elapsed < self.record:
                self.record = elapsed
                self.save_record(elapsed)

    def collides(self, x, y):
        if x < 0 or x >= self.width or y < 0 or y >= self.height:
            return True
        return self.level[y][x] == WALL

    # Вспомогательные для разрешения коллизий
    def get_left_collision(self, x, y):
        # Возвращает x, при котором нет коллизии
        while self.collides(x, y) and x >= 0:
            x -= 1
        return x + 1

    def get_right_collision(self, x, y):
        while self.collides(x, y) and x < self.width:
            x += 1
        return x - 1

    def get_top_collision(self, x, y):
        while self.collides(x, y) and y >= 0:
            y -= 1
        return y + 1

    def get_bottom_collision(self, x, y):
        while self.collides(x, y) and y < self.height:
            y += 1
        return y - 1

    def update(self):
        if self.game_over or self.paused:
            return
        self.update_physics()

    def draw(self):
        os.system('cls' if os.name == 'nt' else 'clear')
        print('═' * self.width)
        elapsed = time.time() - self.start_time
        print(f'  Монеты: {self.coins}/{self.total_coins}   Время: {elapsed:.1f} сек')
        if self.record is not None:
            print(f'  Рекорд: {self.record:.1f} сек')
        print('═' * self.width)

        # Копируем уровень
        grid = [row[:] for row in self.level]
        # Размещаем игрока
        grid[self.player_y][self.player_x] = Fore.GREEN + PLAYER + Style.RESET_ALL
        # Выход
        grid[self.exit_y][self.exit_x] = Fore.YELLOW + EXIT + Style.RESET_ALL
        # Монеты уже раскрашены при выводе
        for row in grid:
            line = ''.join(row)
            # Цвет для монет
            line = line.replace('$', Fore.MAGENTA + '$' + Style.RESET_ALL)
            # Стены
            line = line.replace('#', Fore.BLUE + '#' + Style.RESET_ALL)
            print('│' + line + '│')
        print('═' * self.width)
        status = "ПАУЗА" if self.paused else ("ПОБЕДА!" if self.won else "ИГРА")
        if self.game_over and not self.won:
            status = "ИГРА ОКОНЧЕНА"
        print(f'  {status}  |  ← → - движение  |  Пробел/↑ - прыжок  |  P - пауза  |  R - рестарт  |  Q - выход')

    def handle_input(self):
        while self.running:
            try:
                if keyboard.is_pressed('left') or keyboard.is_pressed('a'):
                    with self.lock:
                        if not self.game_over and not self.paused:
                            self.vx = -1.5
                elif keyboard.is_pressed('right') or keyboard.is_pressed('d'):
                    with self.lock:
                        if not self.game_over and not self.paused:
                            self.vx = 1.5
                else:
                    with self.lock:
                        if not self.game_over and not self.paused:
                            self.vx = 0
                if keyboard.is_pressed('space') or keyboard.is_pressed('up'):
                    with self.lock:
                        self.jump()
                elif keyboard.is_pressed('p'):
                    with self.lock:
                        self.paused = not self.paused
                    time.sleep(0.2)
                elif keyboard.is_pressed('r'):
                    with self.lock:
                        self.reset()
                    time.sleep(0.2)
                elif keyboard.is_pressed('q'):
                    self.running = False
                    break
            except:
                pass
            time.sleep(0.02)

    def reset(self):
        self.player_x, self.player_y = 0, 0
        self.vx = 0
        self.vy = 0
        self.on_ground = False
        self.coins = 0
        self.game_over = False
        self.won = False
        self.paused = False
        self.start_time = time.time()
        # Восстанавливаем уровень
        self.level = [list(row.ljust(WIDTH)) for row in LEVEL]
        self.find_player_and_exit()

    def run(self):
        input_thread = threading.Thread(target=self.handle_input, daemon=True)
        input_thread.start()

        last_update = time.time()
        while self.running:
            now = time.time()
            if now - last_update >= 1.0 / 60:
                with self.lock:
                    self.update()
                    self.draw()
                last_update = now
            time.sleep(0.01)

if __name__ == "__main__":
    game = Platformer()
    try:
        game.run()
    except KeyboardInterrupt:
        print("\nВыход...")

File: platformer.js
// platformer.js - Минимальный платформер на JavaScript (Node.js)
const fs = require('fs');
const keypress = require('keypress');
const readline = require('readline');

const WIDTH = 40;
const HEIGHT = 12;
const GRAVITY = 0.3;
const JUMP_SPEED = -5.5;
const PLAYER = '@';
const WALL = '#';
const COIN = '$';
const EXIT = 'X';
const RECORD_FILE = 'platformer_record.json';

const LEVEL_RAW = [
    '                                        ',
    '                                        ',
    '          #   #                         ',
    '   $      #   #     #   #              ',
    '   @      #   #  $  #   #   X          ',
    '########   ########   #############    ',
];
while (LEVEL_RAW.length < HEIGHT) LEVEL_RAW.push(' '.repeat(WIDTH));
const LEVEL = LEVEL_RAW.map(row => row.padEnd(WIDTH, ' ').split(''));

class Platformer {
    constructor() {
        this.width = WIDTH;
        this.height = HEIGHT;
        this.level = LEVEL.map(row => [...row]);
        this.playerX = 0;
        this.playerY = 0;
        this.vx = 0;
        this.vy = 0;
        this.onGround = false;
        this.coins = 0;
        this.totalCoins = 0;
        this.exitX = 0;
        this.exitY = 0;
        this.findPlayerAndExit();
        this.gameOver = false;
        this.won = false;
        this.paused = false;
        this.running = true;
        this.startTime = Date.now();
        this.record = this.loadRecord();
        this.lastUpdate = Date.now();
        this.timer = null;
    }

    findPlayerAndExit() {
        for (let y = 0; y < this.level.length; y++) {
            for (let x = 0; x < this.level[y].length; x++) {
                const ch = this.level[y][x];
                if (ch === PLAYER) {
                    this.playerX = x;
                    this.playerY = y;
                    this.level[y][x] = ' ';
                } else if (ch === EXIT) {
                    this.exitX = x;
                    this.exitY = y;
                } else if (ch === COIN) {
                    this.totalCoins++;
                }
            }
        }
    }

    loadRecord() {
        try {
            const data = fs.readFileSync(RECORD_FILE, 'utf8');
            return JSON.parse(data).record || null;
        } catch { return null; }
    }

    saveRecord(time) {
        fs.writeFileSync(RECORD_FILE, JSON.stringify({ record: time }));
    }

    jump() {
        if (this.onGround && !this.gameOver && !this.paused) {
            this.vy = JUMP_SPEED;
            this.onGround = false;
        }
    }

    collides(x, y) {
        if (x < 0 || x >= this.width || y < 0 || y >= this.height) return true;
        return this.level[y][x] === WALL;
    }

    getLeftCollision(x, y) {
        while (this.collides(x, y) && x >= 0) x--;
        return x + 1;
    }
    getRightCollision(x, y) {
        while (this.collides(x, y) && x < this.width) x++;
        return x - 1;
    }
    getTopCollision(x, y) {
        while (this.collides(x, y) && y >= 0) y--;
        return y + 1;
    }
    getBottomCollision(x, y) {
        while (this.collides(x, y) && y < this.height) y++;
        return y - 1;
    }

    updatePhysics() {
        this.vy += GRAVITY;
        if (this.vy > 10) this.vy = 10;

        // Горизонталь
        this.playerX += this.vx;
        if (this.collides(this.playerX, this.playerY)) {
            if (this.vx > 0) this.playerX = this.getLeftCollision(this.playerX, this.playerY);
            else if (this.vx < 0) this.playerX = this.getRightCollision(this.playerX, this.playerY);
            this.vx = 0;
        }

        // Вертикаль
        this.playerY += this.vy;
        if (this.collides(this.playerX, this.playerY)) {
            if (this.vy > 0) {
                this.playerY = this.getTopCollision(this.playerX, this.playerY);
                this.vy = 0;
                this.onGround = true;
            } else if (this.vy < 0) {
                this.playerY = this.getBottomCollision(this.playerX, this.playerY);
                this.vy = 0;
            }
        }

        // Монеты
        const cell = this.level[this.playerY]?.[this.playerX];
        if (cell === COIN) {
            this.level[this.playerY][this.playerX] = ' ';
            this.coins++;
        }

        // Выход
        if (this.playerX === this.exitX && this.playerY === this.exitY) {
            this.won = true;
            this.gameOver = true;
            const elapsed = (Date.now() - this.startTime) / 1000;
            if (this.record === null || elapsed < this.record) {
                this.record = elapsed;
                this.saveRecord(elapsed);
            }
        }
    }

    update() {
        if (this.gameOver || this.paused) return;
        this.updatePhysics();
    }

    draw() {
        console.clear();
        const elapsed = (Date.now() - this.startTime) / 1000;
        console.log('═'.repeat(this.width));
        console.log(`  Монеты: ${this.coins}/${this.totalCoins}   Время: ${elapsed.toFixed(1)} сек`);
        if (this.record !== null) console.log(`  Рекорд: ${this.record.toFixed(1)} сек`);
        console.log('═'.repeat(this.width));

        const grid = this.level.map(row => [...row]);
        grid[this.playerY][this.playerX] = PLAYER;
        grid[this.exitY][this.exitX] = EXIT;

        for (const row of grid) {
            let line = row.join('');
            line = line.replace(/\$/g, '\x1b[35m$\x1b[0m');
            line = line.replace(/#/g, '\x1b[34m#\x1b[0m');
            console.log('│' + line + '│');
        }
        console.log('═'.repeat(this.width));
        let status = this.paused ? 'ПАУЗА' : this.won ? 'ПОБЕДА!' : 'ИГРА';
        if (this.gameOver && !this.won) status = 'ИГРА ОКОНЧЕНА';
        console.log(`  ${status}  |  ← → - движение  |  Пробел/↑ - прыжок  |  P - пауза  |  R - рестарт  |  Q - выход`);
    }

    reset() {
        this.playerX = 0;
        this.playerY = 0;
        this.vx = 0;
        this.vy = 0;
        this.onGround = false;
        this.coins = 0;
        this.gameOver = false;
        this.won = false;
        this.paused = false;
        this.startTime = Date.now();
        this.level = LEVEL.map(row => [...row]);
        this.findPlayerAndExit();
    }

    handleKey(ch, key) {
        if (!key) return;
        if (key.name === 'left' || key.name === 'a') {
            if (!this.gameOver && !this.paused) this.vx = -1.5;
        } else if (key.name === 'right' || key.name === 'd') {
            if (!this.gameOver && !this.paused) this.vx = 1.5;
        } else if (key.name === 'space' || key.name === 'up') {
            this.jump();
        } else if (key.name === 'p') {
            this.paused = !this.paused;
        } else if (key.name === 'r') {
            this.reset();
        } else if (key.name === 'q') {
            this.running = false;
            process.stdin.pause();
            process.exit(0);
        }
        // Остановка движения при отпускании клавиш
        // Но keypress не даёт события отпускания, поэтому сделаем отдельный таймер
    }

    run() {
        keypress(process.stdin);
        process.stdin.setRawMode(true);
        process.stdin.resume();
        process.stdin.on('keypress', (ch, key) => this.handleKey(ch, key));

        // Таймер для сброса vx при отпускании (упрощённо)
        setInterval(() => {
            if (!this.gameOver && !this.paused) {
                // Если не нажата ни одна клавиша, vx=0 (но мы не знаем)
                // Вместо этого используем событие keypress для установки vx
                // и отдельно будем сбрасывать при отсутствии нажатий.
                // Для простоты оставим vx постоянным при удержании.
            }
        }, 50);

        const gameLoop = () => {
            if (!this.running) return;
            const now = Date.now();
            if (now - this.lastUpdate >= 1000 / 60) {
                this.update();
                this.draw();
                this.lastUpdate = now;
            }
            this.timer = setTimeout(gameLoop, 16);
        };
        gameLoop();
    }
}

const game = new Platformer();
game.run();
